### PR TITLE
read only query

### PR DESCRIPTION
**Pull request by DvirDukhan**

### CODE REVIEW

File: src/module.c [hunk 1]
```diff
@@ -142,6 +142,11 @@ int RedisModule_OnLoad(RedisModuleCtx *ctx, RedisModuleString **argv, int argc)
 		return REDISMODULE_ERR;
 	}
 
+	if(RedisModule_CreateCommand(ctx, "graph.RO_QUERY", CommandDispatch, "readonly", 1, 1,
```
Comment (jeffreylovitz) [inline src/module.c:145]: I think the flags should be `readonly deny-oom`.

File: tests/flow/test_ro_query.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        try:
+            # Every normal command is write command, see that replica connection throws an exception.
+            raw_result_set = slave_con.execute_command("GRAPH.QUERY", graph_name, "MATCH (n) RETURN COUNT(n)", "--compact")
+            result_set = query_result.QueryResult(graph, raw_result_set).result_set
+            assert(False)
+        except:
+            # Expecting an error.
+            pass
+        # Close replica to avoid memcheck
+        slave_con.shutdown()
```
Comment (jeffreylovitz) [inline tests/flow/test_ro_query.py:68]: This confuses my local test run:
```
        test_ro_query:test_read_only_query.test02_test_replica_read_only                         
        [PASS]                                                                               
        process is not alive, might have crash durring test execution, check this out. server id : 2
        check the redis log at: ./logs/ba3a69096abe47ed8e836f6612971e8a.slave-2-test_ro_query.__init__-oss.log
        Printing only REDIS BUG REPORT START and STACK TRACE         
```
Why is it necessary?

**Review by jeffreylovitz — CHANGES_REQUESTED** [review]
This looks good to me outside of the CI failure, a few minor notes! Also, I believe this should be documented.

File: src/module.c [hunk 3]
```diff
@@ -126,7 +126,7 @@ int RedisModule_OnLoad(RedisModuleCtx *ctx, RedisModuleString **argv, int argc)
 	RedisModule_Log(ctx, "notice", "Thread pool created, using %d threads.", threadCount);
 
 	// Initialize array of command contexts
-	command_ctxs = calloc(threadCount + 1, sizeof(CommandCtx*));
+	command_ctxs = calloc(threadCount + 1, sizeof(CommandCtx *));
```
Comment (swilly22) [inline src/module.c:129]: Why the extra space in `sizeof(CommandCtx *));`
I prefer it to be `sizeof(CommandCtx*));`

File: src/module.c (hunk 1 above)
Comment (swilly22) [reply src/module.c:145]: From [redis module api](https://redis.io/topics/modules-api-ref)
Any idea if the following might be relevant ?
```
"allow-stale": The command is allowed to run on slaves that don't serve stale data. Don't use if you don't know what this means.
```

File: src/commands/cmd_query.c [hunk 4]
```diff
@@ -142,6 +142,11 @@ void Graph_Query(void *args) {
 	if(exec_type == EXECUTION_TYPE_INVALID) goto cleanup;
 
 	bool readonly = AST_ReadOnly(ast->root);
+	if(!readonly && strcasecmp(CommandCtx_GetCommandName(command_ctx), "graph.RO_QUERY") == 0) {
```
Comment (swilly22) [inline src/commands/cmd_query.c:145]: Might be a bit cleaner if `strcasecmp(CommandCtx_GetCommandName(command_ctx), "graph.RO_QUERY") == 0`
was an inline function, e.g. `inline bool _readonly_cmd_mode(CommandCtx *ctx)`

File: tests/flow/test_ro_query.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+import os
+import sys
+import time
+from RLTest import Env
+from redisgraph import Graph, Node, Edge, query_result
+
+master_con = None
+slave_con = None
+
+
```
Comment (swilly22) [inline tests/flow/test_ro_query.py:11]: Remove extra line

File: tests/flow/test_ro_query.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+class test_read_only_query(FlowTestsBase):
+    def __init__(self):
+        self.env = Env(moduleArgs='VKEY_MAX_ENTITY_COUNT 3', useSlaves=True)
```
Comment (swilly22) [inline tests/flow/test_ro_query.py:14]: Why do we need to set `VKEY_MAX_ENTITY_COUNT'` ?

File: tests/flow/test_ro_query.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+        global master_con
+        global slave_con
```
Comment (swilly22) [inline tests/flow/test_ro_query.py:16]: Swap
```
global slave_con
global master_con
```
order by line length.

File: tests/flow/test_ro_query.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+        master_con = self.env.getConnection()
+        slave_con = self.env.getSlaveConnection()
+
+    def test01_test_simple_read_only_command(self):
+        graph_name = "Test_RO_QUERY_command"
```
Comment (swilly22) [inline tests/flow/test_ro_query.py:21]: Add a comment explaining the purpose of this test function.

File: tests/flow/test_ro_query.py [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+        except:
+            # Expecting an error.
+            pass
+        try:
+            raw_result_set = master_con.execute_command("GRAPH.RO_QUERY", graph_name, "MATCH(n) DELETE n", "--compact")
+            result_set = query_result.QueryResult(graph, raw_result_set).result_set
+            assert(False)
+        except:
+            # Expecting an error.
+            pass
```
Comment (swilly22) [inline tests/flow/test_ro_query.py:48]: I think a single write query is enough, we might want to add a different unit test which checks AST readonly function.
might want to check write procedures e.g. index creation / deletion.

File: tests/flow/test_ro_query.py [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+
+    def test02_test_replica_read_only(self):
+        graph_name = "Test_RO_QUERY_command_on_replica"
+        graph = Graph(graph_name, master_con)
+        graph.query("UNWIND range(0,20) as i CREATE ()")
+        slave_con.execute_command("REPLICAOF", "localhost", "6379")
```
Comment (swilly22) [inline tests/flow/test_ro_query.py:54]: Can't we get host port information from the `master_con` object?

File: tests/flow/test_ro_query.py [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+        time.sleep(2)
```
Comment (swilly22) [inline tests/flow/test_ro_query.py:55]: On way of getting indication from `slave_con` to know if it is sync?

**Review by swilly22 — CHANGES_REQUESTED** [review]

File: tests/flow/test_ro_query.py (hunk 2 above)
Comment (DvirDukhan) [reply tests/flow/test_ro_query.py]: removed

File: tests/flow/test_ro_query.py (hunk 10 above)
Comment (DvirDukhan) [reply tests/flow/test_ro_query.py:56]: unfortunately, this is not exposed in `redis-py` connection API

File: tests/flow/test_ro_query.py (hunk 11 above)
Comment (DvirDukhan) [reply tests/flow/test_ro_query.py:57]: Added from RediSearch, I don't know if it is cleaner

File: tests/flow/test_ro_query.py [hunk 12]
```diff
@@ -2,22 +2,54 @@
...
+def checkSlaveSynced(env, slaveConn, command, expected_result, time_out=5):
+    try:
+        with TimeLimit(time_out):
+            res = slaveConn.execute_command(*command)
+            while res != expected_result:
+                res = slaveConn.execute_command(*command)
+    except TimeoutException:
+        env.assertTrue(False, message='Failed waiting for command to be executed on slave')
+    except Exception as e:
+        env.assertTrue(False, message=e.message)
```
Comment (swilly22) [inline tests/flow/test_ro_query.py:40]: I think this is too complex, let's discus alternatives.

**Review by swilly22 — CHANGES_REQUESTED** [review]

File: tests/flow/test_ro_query.py [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+    def __init__(self, timeout):
+        self.timeout = timeout
+    def __enter__(self):
+        signal.signal(signal.SIGALRM, self.handler)
+        signal.setitimer(signal.ITIMER_REAL, self.timeout, 0)
+    def __exit__(self, exc_type, exc_value, traceback):
+        signal.setitimer(signal.ITIMER_REAL, 0)
+        signal.signal(signal.SIGALRM, signal.SIG_DFL)
+    def handler(self, signum, frame):
+        raise TimeoutException()
```
Comment (swilly22) [inline tests/flow/test_ro_query.py:29]: Replace with a simple call to `Sleep`

**Review by swilly22 — CHANGES_REQUESTED** [review]

**Review by swilly22 — APPROVED** [review]

**Review by swilly22 — APPROVED** [review]

**Comment by swilly22** [comment]
Resolves #1071

**Pull request merged**